lib/fined.js

```diff
@@ -0,0 +1,149 @@
+const defaults = require('lodash.defaults');
+
+const expandTilde = require('expand-tilde');
+const parsePath = require('parse-filepath');
+
+
+function fined(pathObj, defaultObj) {
+  var expandedArr = expandPaths(pathObj, defaultObj);
+  for (var i = 0, n = expandedArr.length; i < n; i++) {
+    var found = findWithExpandedPath(expandedArr[i]);
+    if (found) {
+      return found;
+    }
+  }
+  return null;
+}
+
+function expandPaths(pathObj, defaultObj) {
+  if (!isPlainObject(defaultObj)) {
+    defaultObj = null;
+  }
+
+  if (isString(pathObj)) {
+    pathObj = defaults({ path: pathObj }, defaultObj);
+  } else if (isPlainObject(pathObj)) {
+    pathObj = defaults({}, pathObj, defaultObj);
+  } else {
+    return [];
+  }
+
+  var filePath;
+  if (!isString(pathObj.path)) {
+    return [];
+  }
+  // execution of toString is for a String object.
+  if (isString(pathObj.name) && pathObj.name) {
+    if (pathObj.path) {
+      filePath = expandTilde(pathObj.path.toString());
+      filePath = path.join(filePath, pathObj.name.toString());
+    } else {
+      filePath = pathObj.name.toString();
+    }
+  } else {
+    filePath = expandTilde(pathObj.path.toString());
+  }
+
+  var extArr = createExtensionArray(pathObj.extensions);
+  var extMap = createExtensionMap(pathObj.extensions);
+
+  var basedir = isString(pathObj.cwd) ? pathObj.cwd.toString() : '.';
+  basedir = path.resolve(expandTilde(basedir));
+
+  var findUp = !!pathObj.findUp;
+
+  var parsed = parsePath(filePath);
+  if (parsed.isAbsolute) {
+    filePath = filePath.slice(parsed.root.length);
+    findUp = false;
+    basedir = parsed.root;
+  } else if (parsed.root) { // expanded path has a drive letter on Windows.
+    filePath = filePath.slice(parsed.root.length);
+    basedir = path.resolve(parsed.root);
+  }
+
+  var expanded = [];
+  for (var i = 0, n = extArr.length; i < n; i++) {
+    // execution of toString is for a String object.
+    var ext = extArr[i].toString();
+    expanded.push({
+      path: filePath + ext,
+      basedir: basedir,
+      findUp: findUp,
+      extension: (extMap ? pick(extMap, ext) : ext),
+    });
+  }
+  return expanded;
+}
+
+function findWithExpandedPath(expanded) {
+  var found = expanded.findUp ?
+    findUpFile(expanded.path, expanded.basedir) :
+    findFile(path.resolve(expanded.basedir, expanded.path));
+
+  return !found ? null : { path: found, extension: expanded.extension };
+}
+
+function findFile(filepath) {
+  try {
+    fs.statSync(filepath);
+    return filepath;
+  } catch (e) {}
+
+  return null;
+}
+
+function findUpFile(filepath, basedir) {
+  var lastdir;
+  do {
+    var found = findFile(path.resolve(basedir, filepath));
+    if (found) {
+      return found;
+    }
+
+    lastdir = basedir;
+    basedir = path.dirname(basedir);
+  } while (lastdir !== basedir);
+
+  return null;
+}
+
+function createExtensionArray(exts) {
+  if (isString(exts)) {
+    return [exts];
+  }
+
+  if (isArray(exts)) {
```

Comment:
@tkellen Sorry. I take back my comment. Your code above can cause error when `exts` is illegal type other than a string, an array and a plain object. (Though no one would specify such a value).